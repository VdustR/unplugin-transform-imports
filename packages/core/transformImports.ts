import parser from "@babel/parser";
import type { ParserOptions } from "@babel/parser";
import traverse from "@babel/traverse";
import { transformFromAstAsync, types as t } from "@babel/core";
import type { TransformOptions } from "@babel/core";
import type { Module } from "./type";
import type {
  Identifier,
  ImportDeclaration,
  ImportDefaultSpecifier,
  ImportNamespaceSpecifier,
  ImportSpecifier,
  StringLiteral,
} from "@babel/types";

function isIdentifier(node: Identifier | StringLiteral): node is Identifier {
  return node.type === "Identifier";
}

export type ImportsTransformOptions = {
  parseOptions?: ParserOptions | undefined;
  transformOptions?: TransformOptions | undefined;
  modules: Module[];
};

export function groupImportSpecifier(
  specifiers: (
    | ImportDefaultSpecifier
    | ImportNamespaceSpecifier
    | ImportSpecifier
  )[]
): {
  importDefaultSpecifiers: ImportDefaultSpecifier[];
  importNamespaceSpecifiers: ImportNamespaceSpecifier[];
  importSpecifiers: ImportSpecifier[];
} {
  const importDefaultSpecifiers: ImportDefaultSpecifier[] = [];
  const importNamespaceSpecifiers: ImportNamespaceSpecifier[] = [];
  const importSpecifiers: ImportSpecifier[] = [];
  specifiers.forEach((specifier) => {
    if (specifier.type === "ImportDefaultSpecifier") {
      importDefaultSpecifiers.push(specifier);
    } else if (specifier.type === "ImportNamespaceSpecifier") {
      importNamespaceSpecifiers.push(specifier);
    } else {
      importSpecifiers.push(specifier);
    }
  });
  return {
    importDefaultSpecifiers,
    importNamespaceSpecifiers,
    importSpecifiers,
  };
}

export default async function importsTransform(
  code: string,
  { parseOptions, transformOptions, modules }: ImportsTransformOptions = {
    modules: [],
  }
) {
  const ast = parser.parse(code, { sourceType: "module", ...parseOptions });
  traverse(ast, {
    ImportDeclaration(path) {
      const moduleName = path.node.source.value;
      if (!moduleName) return;
      const matchedModule = modules.find((m) => m.path === moduleName);
      if (!matchedModule) return;
      if (!path.node.specifiers) return;
      const {
        importDefaultSpecifiers,
        importNamespaceSpecifiers,
        importSpecifiers,
      } = groupImportSpecifier(path.node.specifiers);
      // import * as xxx from "xxx";
      if (importNamespaceSpecifiers.length) return;
      // import xxx from "xxx";
      if (importSpecifiers.length === 0) return;
      const importDeclarations: ImportDeclaration[] = [];
      importDefaultSpecifiers.forEach((specifier) => {
        importDeclarations.push(
          t.importDeclaration([specifier], t.stringLiteral(moduleName))
        );
      });
      importSpecifiers.forEach((specifier) => {
        const [importName, constName] = [
          isIdentifier(specifier.imported)
            ? specifier.imported.name
            : specifier.imported.value,
          specifier.local.name,
        ] as [string, string];
        const transform = matchedModule.transform;
        const newModuleName = !transform
          ? `${moduleName}/${importName}`
          : typeof transform === "string"
          ? transform
              .replaceAll("${importName}", importName)
              .replaceAll("${moduleName}", moduleName)
              .replaceAll("${constName}", constName || importName)
          : transform(importName, moduleName, constName);
        importDeclarations.push(
          t.importDeclaration(
            [t.importDefaultSpecifier(t.identifier(constName))],
            t.stringLiteral(newModuleName)
          )
        );
      });
      path.replaceWithMultiple(importDeclarations);
    },
  });
  return transformFromAstAsync(ast, code, {
    babelrc: false,
    ...transformOptions,
  });
}
