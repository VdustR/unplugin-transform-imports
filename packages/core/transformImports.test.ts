import transformImports from "./transformImports";
import { expect, it } from "vitest";
import { resolve } from "path";
import { readFileSync } from "fs-extra";

const code = readFileSync(resolve(__dirname, "test/index.tsx"), "utf8");

it("Test build", async () => {
  const babelFileResult = await transformImports(code, {
    parseOptions: {
      plugins: ["typescript", "jsx"],
    },
    modules: [
      { path: "foo" },
      {
        path: "@bar/core",
        transform: "@bar/core/${importName}",
      },
      {
        path: "baz",
        transform: `\${moduleName}/\${importName}/\${importName}/deep/\${constName}/\${constName}`,
      },
      {
        path: "@qux/core",
        transform: "@qux/core/${constName}",
      },
      {
        path: "quux",
        transform: (importName) => `quux/dist/${importName}`,
      },
      {
        path: "@corge/react",
        transform: (importName, moduleName, constName) =>
          `${moduleName}/dist/${importName}/${importName}/deep/${constName}/${constName}`,
      },
      {
        path: "grault-vue",
        transform: (_importName, _moduleName, constName) =>
          `grault-vue/dist/${constName}`,
      },
    ],
  });
  expect(babelFileResult?.code).toMatchSnapshot();
});
