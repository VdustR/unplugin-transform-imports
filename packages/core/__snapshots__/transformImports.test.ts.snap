// Vitest Snapshot v1

exports[`Test build 1`] = `
"// @ts-ignore
import foo from \\"foo/foo\\";
import foo1 from \\"foo/foo1Ex\\";
import foo2 from \\"foo/foo2\\"; // @ts-ignore

import bar from \\"@bar/core/bar\\";
import bar1 from \\"@bar/core/bar1Ex\\";
import bar2 from \\"@bar/core/bar2\\";
import // tailing comment
quxDefault from \\"baz\\";
import baz from \\"baz/baz/baz/deep/baz/baz\\";
import baz1 from \\"baz/baz1Ex/baz1Ex/deep/baz1/baz1\\";
import baz2 from \\"baz/baz2/baz2/deep/baz2/baz2\\";
import qux from \\"@qux/core/qux\\";
import qux1 from \\"@qux/core/qux1\\";
import qux2 from \\"@qux/core/qux2\\"; // @ts-ignore

import * as quxAll from \\"@qux/core\\";
import quux from \\"quux/dist/default\\";
import quux1 from \\"quux/dist/quux1Ex\\";
import quux2 from \\"quux/dist/quux2\\"; // @ts-ignore

import corge from \\"@corge/react/dist/corge/corge/deep/corge/corge\\";
import corge1 from \\"@corge/react/dist/corge1Ex/corge1Ex/deep/corge1/corge1\\";
import corge2 from \\"@corge/react/dist/corge2/corge2/deep/corge2/corge2\\"; // @ts-ignore

import grault from \\"grault-vue/dist/grault\\";
import grault1 from \\"grault-vue/dist/grault1\\";
import grault2 from \\"grault-vue/dist/grault2\\";
const str = (\\"str\\" as const);
const component = <div />;
console.log({
  foo,
  foo1,
  foo2,
  bar,
  bar1,
  bar2,
  baz,
  baz1,
  baz2,
  quxDefault,
  qux,
  qux1,
  qux2,
  quxAll,
  quux,
  quux1,
  quux2,
  corge,
  corge1,
  corge2,
  grault,
  grault1,
  grault2,
  str,
  component
});"
`;
